web-api/src/mind.rs

```diff
@@ -402,49 +404,52 @@ async fn run_persona_benchmark() -> Result<(), Error> {
 }
 
 /// Runs the user-based mind benchmark
-#[test]
+#[tokio::test]
 #[ignore]
-fn run_user_benchmark() -> Result<(), Error> {
+async fn run_user_benchmark() -> Result<(), Error> {
     let document_provider = DocumentProvider::new("news.tsv")?;
 
-    let impressions = read("behaviors.tsv")?;
+    let state = State::new(Storage::default()).unwrap();
+    state
+        .insert(document_provider.to_documents())
+        .await
+        .unwrap();
 
     let nranks = vec![3];
     let mut ndcgs = Array::zeros((nranks.len(), 0));
 
     // Loop over all impressions, prepare reranker with news in click history
     // and rerank the news in an impression
-    for impression in impressions {
+    for impression in read("behaviors.tsv")? {
         let impression: Impression = impression?;
+        let user = UserId::new(impression.user_id).unwrap();
 
-        // Placeholder for interacting with the entire click history
-        for click in impression.clicks {
-            match document_provider.get(&click) {
-                Some(document) => println!("The document {:?} was interacted.", document),
-                None => println!("Document id {} not found.", click),
-            }
-        }
+        state.interact(&user, &impression.clicks).await.unwrap();
```

Comment:
The history of clicks of a single user does not change between his/her different impressions, which means we could interact with clicked news only once e.g when we encounter a new user id. We could use a set of user ids for this

Reply:
True, we could do this. However, we also need to maintain the state of the user id and keep track of whether the user id was already interacted with. Would mark this as further optimization as long as we don't have any need in increasing the computational speed. In the worst case, it takes around 1.5 times longer than before.

Reply:
I think it is not only a speed issue, but also interacting with the same articles in every impression may change the results as the weights of topics are incremented multiple times and also CoIs are changed